packages/vite/src/node/utils.ts

```diff
@@ -622,6 +623,20 @@ export function copyDir(srcDir: string, destDir: string): void {
   }
 }
 
+export async function recursiveReaddir(dir: string): Promise<string[]> {
+  if (!fs.existsSync(dir)) {
+    return []
+  }
+  const dirents = await fsp.readdir(dir, { withFileTypes: true })
+  const files = await Promise.all(
+    dirents.map((dirent) => {
+      const res = path.resolve(dir, dirent.name)
+      return dirent.isDirectory() ? recursiveReaddir(res) : normalizePath(res)
+    }),
+  )
+  return Array.prototype.concat(...files)
+}
```

Comment:
I think we need to ignore permission errors here.

Reply:
done at [fix: handle EACCES in recursiveReaddir](https://github.com/vitejs/vite/pull/15195/commits/ee1d6a2be3b5f22c64e2999a790ba5e346dfdaad)